README.md
# Flutter Screen Lock

This Flutter plugin provides an feature for screen lock.
Enter your passcode to unlock the screen.
You can also use biometric authentication as an option.

<img src="https://raw.githubusercontent.com/naoki0719/flutter_screen_lock/master/resources/flutter_screen_lock.gif" />

## Features

- Any number of digits can be specified
- You can change `Cancel` and `Delete` text
- The UI expands and contracts according to the size of the device
- You can disable cancellation
- You can use biometrics
- Biometrics can be displayed on first launch
- Unlocked callback
- Limit the maximum number of retries

## Pending

- Change the text color of Cancel, Delete, and Biometric

## Usage

You can easily lock the screen with the following code.  
To unlock, enter correctString.

### Simple

If the passcode you entered matches, you can callback onUnlocked.

```dart
import 'package:flutter_screen_lock/flutter_screen_lock.dart';

showLockScreen(
  context: context,
  correctString: '1234',
  onUnlocked: () => print('Unlocked.'),
);
```

### Change digits

Default 4 digits can be changed. Change the correctString accordingly.

```dart
import 'package:flutter_screen_lock/flutter_screen_lock.dart';

showLockScreen(
  context: context,
  digits: 6,
  correctString: '123456',
);
```

### Use local_auth

Specify `canBiometric` and `biometricAuthenticate`.

Add local_auth processing to `biometricAuthenticate`. See the following page for details.

https://pub.dev/packages/local_auth

```dart
import 'package:flutter_screen_lock/flutter_screen_lock.dart';

showLockScreen(
  context: context,
  correctString: '1234',
  canBiometric: true,
  // biometricButton is default Icon(Icons.fingerprint)
  // When you want to change the icon with `BiometricType.face`, etc.
  biometricButton: Icon(Icons.face),
  biometricAuthenticate: (context) async {
    final localAuth = LocalAuthentication();
    final didAuthenticate =
        await localAuth.authenticateWithBiometrics(
            localizedReason: 'Please authenticate');

    if (didAuthenticate) {
      return true;
    }

    return false;
  },
);
```

### Open biometric first & onUnlocked callback

add option showBiometricFirst.

```dart
showLockScreen(
  context: context,
  correctString: '1234',
  canBiometric: true,
  showBiometricFirst: true,
  biometricAuthenticate: (context) async {
    final localAuth = LocalAuthentication();
    final didAuthenticate =
        await localAuth.authenticateWithBiometrics(
            localizedReason: 'Please authenticate');

    if (didAuthenticate) {
      return true;
    }

    return false;
  },
  onUnlocked: () {
    print('Unlocked.');
  },
);
```

### Can't cancel

This is the case where you want to force authentication when the app is first launched.

```dart
showLockScreen(
  context: context,
  correctString: '1234',
  canCancel: false,
);
```

### Customize text

You can change `Cancel` and `Delete` text.
We recommend no more than 6 characters at this time.

```dart
showLockScreen(
  context: context,
  correctString: '1234',
  cancelText: 'Close',
  deleteText: 'Remove',
);
```

### Verifycation passcode (v1.1.1)

use `showConfirmPasscode` function.

<img src="https://raw.githubusercontent.com/naoki0719/flutter_screen_lock/master/resources/flutter_screen_lock_confirm.gif" />

```dart
showConfirmPasscode(
  context: context,
  confirmTitle: 'This is the second input.',
  onCompleted: (context, verifyCode) {
    // verifyCode is verified passcode
    print(verifyCode);
    // Please close yourself
    Navigator.of(context).maybePop();
  },
)
```

### Customize your style (v1.1.2)

use `circleInputButtonConfig` option.

<img src="https://raw.githubusercontent.com/naoki0719/flutter_screen_lock/master/resources/customize_styles.png" />

```dart
showLockScreen(
  context: context,
  correctString: '1234',
  backgroundColor: Colors.grey.shade50,
  backgroundColorOpacity: 1,
  circleInputButtonConfig: CircleInputButtonConfig(
    textStyle: TextStyle(
      fontSize: MediaQuery.of(context).size.width * 0.1,
      color: Colors.white,
    ),
    backgroundColor: Colors.blue,
    backgroundOpacity: 0.5,
    shape: RoundedRectangleBorder(
      side: BorderSide(
        width: 1,
        color: Colors.blue,
        style: BorderStyle.solid,
      ),
    ),
  ),
)
```

### Max retries / Incorrect event (v1.2.8)

```dart
showLockScreen(
  context: context,
  correctString: '1234',
  maxRetries: 1, // -1 is unlimited
  onError: (retries) {
    print(retries);
  },
  didMaxRetries: () {
    Navigator.pop(context);
    showAboutDialog(
      context: context,
      children: [
        Text('The maximum number of retries has been reached.'),
      ],
    );
  },
  biometricAuthenticate: (_) async {
    final localAuth = LocalAuthentication();
    final didAuthenticate = await localAuth.authenticate(
      localizedReason: 'Please authenticate',
      biometricOnly: true,
    );

    if (didAuthenticate) {
      return true;
    }

    return false;
  },
  canBiometric: true,
)
```

## Help

### How to prevent the background from being transparent

Set the `backgroundColorOpacity` option to 1

## Apps I use

TimeKey

[iOS](https://apps.apple.com/us/app/timekey-authenticator/id1506129753)

[Android](https://play.google.com/store/apps/details?id=net.incrementleaf.TimeKey)


CHANGELOG.md
# Changelog

## [2.0.3-nullsafety] - 2021-02-22

- `didMaxRetries` and `onError` are not handled by biometric.

## [1.2.9] - 2021-02-22

- `didMaxRetries` and `onError` are not handled by biometric.

## [2.0.2-nullsafety] - 2021-02-22

- Add the `onError` event when input fails.

## [1.2.8+1] - 2021-02-22

- Fixed the README.

## [1.2.8] - 2021-02-22

- Add the `onError` event when input fails.

## [1.2.7+1] - 2021-02-22

- Added the ability to limit the maximum retries.

## [2.0.1-nullsafety] - 2021-02-22

- Added the ability to limit the maximum retries.

## [1.2.7] - 2021-02-22

- Added the ability to limit the maximum retries.

## [2.0.0-nullsafety] - 2021-02-17

- Supports null safety

## [1.2.6] - 2021-01-03

- Fixed to pop after biometric authentication and then call the unlocked function.

## [1.2.5] - 2021-01-02

- Fixed to center the screen content.

## [1.2.4] - 2020-07-29

- Fixed an exception when deleting input.

## [1.2.3] - 2020-07-29

- Fix a missing exception in onUnlocked.

## [1.2.2] - 2020-07-28

- Fix exception for onUnlocked.

## [1.2.1] - 2020-07-08

- Added custom biometric button.

## [1.2.0] - 2020-07-08

- Added unlocked callback function.
- Deprecate biometricFunction. Instead, we have a biometricAuthenticate.
- If you return true with biometricAuthenticate, you can set onUnlocked to Callback. Also, Navigation.pop is now automatic.

## [1.1.2] - 2020-05-10

- Added the ability to customize buttons and backgrounds.

## [1.1.1] - 2020-05-10

- Fix to call showBiometricFunction when the animation is complete

## [1.1.0+1] - 2020-03-03

- Added video for confirm screen

## [1.1.0] - 2020-02-28

- New feature in verification passcode

## [1.0.0+2] - 2020-02-18

- Change example readme

## [1.0.0+1] - 2020-02-17

- Fix readme

## [1.0.0] - 2020-02-16

- First release

### Features

- Any number of digits can be specified
- You can change `Cancel` and `Delete` text
- The UI expands and contracts according to the size of the device
- You can disable cancellation
- You can use biometrics
- Biometrics can be displayed on first launch
- Verification passcode
- Unlocked callback


lib/lock_screen.dart
import 'dart:async';
import 'dart:ui';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/services.dart';
import 'dot_secret_ui.dart';
import 'circle_input_button.dart';

Future showConfirmPasscode({
  required BuildContext context,
  String title = 'Please enter passcode.',
  String confirmTitle = 'Please enter confirm passcode.',
  String cancelText = 'Cancel',
  String deleteText = 'Delete',
  int digits = 4,
  DotSecretConfig dotSecretConfig = const DotSecretConfig(),
  void Function(BuildContext, String)? onCompleted,
  Color backgroundColor = Colors.white,
  double backgroundColorOpacity = 0.5,
  CircleInputButtonConfig circleInputButtonConfig =
      const CircleInputButtonConfig(),
  int maxRetries = 1,
  void Function()? didMaxRetries,
  void Function(int retries)? onError,
}) {
  return Navigator.of(context).push(
    PageRouteBuilder(
      opaque: false,
      pageBuilder: (
        BuildContext context,
        Animation<double> animation,
        Animation<double> secodaryAnimation,
      ) {
        return LockScreen(
          title: title,
          confirmTitle: confirmTitle,
          confirmMode: true,
          digits: digits,
          dotSecretConfig: dotSecretConfig,
          onCompleted: onCompleted,
          cancelText: cancelText,
          deleteText: deleteText,
          backgroundColor: backgroundColor,
          backgroundColorOpacity: backgroundColorOpacity,
          circleInputButtonConfig: circleInputButtonConfig,
          maxRetries: maxRetries,
          didMaxRetries: didMaxRetries,
          onError: onError,
        );
      },
      transitionsBuilder: (
        BuildContext context,
        Animation<double> animation,
        Animation<double> secondaryAnimation,
        Widget child,
      ) {
        return SlideTransition(
          position: Tween<Offset>(
            begin: const Offset(0.0, 2.4),
            end: Offset.zero,
          ).animate(animation),
          child: SlideTransition(
            position: Tween<Offset>(
              begin: Offset.zero,
              end: const Offset(0.0, 2.4),
            ).animate(secondaryAnimation),
            child: child,
          ),
        );
      },
    ),
  );
}

Future showLockScreen({
  required BuildContext context,
  String correctString = '',
  String title = 'Please enter passcode.',
  String cancelText = 'Cancel',
  String deleteText = 'Delete',
  int digits = 4,
  DotSecretConfig dotSecretConfig = const DotSecretConfig(),
  bool canCancel = true,
  void Function(BuildContext, String)? onCompleted,
  Widget biometricButton = const Icon(Icons.fingerprint),
  bool canBiometric = false,
  bool showBiometricFirst = false,
  Future<bool> Function(BuildContext)? biometricAuthenticate,
  Color backgroundColor = Colors.white,
  double backgroundColorOpacity = 0.5,
  CircleInputButtonConfig circleInputButtonConfig =
      const CircleInputButtonConfig(),
  void Function()? onUnlocked,
  int maxRetries = 1,
  void Function()? didMaxRetries,
  void Function(int retries)? onError,
}) {
  return Navigator.of(context).push(
    PageRouteBuilder(
      opaque: false,
      pageBuilder: (
        BuildContext context,
        Animation<double> animation,
        Animation<double> secodaryAnimation,
      ) {
        var _showBiometricFirstController = StreamController<void>();

        animation.addStatusListener((status) {
          // Calling the biometric on completion of the animation.
          if (status == AnimationStatus.completed) {
            _showBiometricFirstController.add(null);
          }
        });

        return LockScreen(
          correctString: correctString,
          title: title,
          digits: digits,
          dotSecretConfig: dotSecretConfig,
          onCompleted: onCompleted,
          canCancel: canCancel,
          cancelText: cancelText,
          deleteText: deleteText,
          biometricButton: biometricButton,
          canBiometric: canBiometric,
          showBiometricFirst: showBiometricFirst,
          showBiometricFirstController: _showBiometricFirstController,
          biometricAuthenticate: biometricAuthenticate,
          backgroundColor: backgroundColor,
          backgroundColorOpacity: backgroundColorOpacity,
          circleInputButtonConfig: circleInputButtonConfig,
          onUnlocked: onUnlocked,
          maxRetries: maxRetries,
          didMaxRetries: didMaxRetries,
          onError: onError,
        );
      },
      transitionsBuilder: (
        BuildContext context,
        Animation<double> animation,
        Animation<double> secondaryAnimation,
        Widget child,
      ) {
        return SlideTransition(
          position: Tween<Offset>(
            begin: const Offset(0.0, 2.4),
            end: Offset.zero,
          ).animate(animation),
          child: SlideTransition(
            position: Tween<Offset>(
              begin: Offset.zero,
              end: const Offset(0.0, 2.4),
            ).animate(secondaryAnimation),
            child: child,
          ),
        );
      },
    ),
  );
}

class LockScreen extends StatefulWidget {
  final String? correctString;
  final String title;
  final String confirmTitle;
  final bool confirmMode;
  final Widget? rightSideButton;
  final int digits;
  final DotSecretConfig dotSecretConfig;
  final CircleInputButtonConfig circleInputButtonConfig;
  final bool canCancel;
  final String? cancelText;
  final String? deleteText;
  final Widget biometricButton;
  final void Function(BuildContext, String)? onCompleted;
  final bool canBiometric;
  final bool showBiometricFirst;
  final Future<bool> Function(BuildContext)? biometricAuthenticate;
  final StreamController<void>? showBiometricFirstController;
  final Color backgroundColor;
  final double backgroundColorOpacity;
  final void Function()? onUnlocked;

  /// -1 is unlimited. [Default -1]
  final int maxRetries;
  final void Function()? didMaxRetries;

  /// Called every time correctString or biometric fails.
  final void Function(int retries)? onError;

  LockScreen({
    this.correctString,
    this.title = 'Please enter passcode.',
    this.confirmTitle = 'Please enter confirm passcode.',
    this.confirmMode = false,
    this.digits = 4,
    this.dotSecretConfig = const DotSecretConfig(),
    this.circleInputButtonConfig = const CircleInputButtonConfig(),
    this.rightSideButton,
    this.canCancel = true,
    this.cancelText,
    this.deleteText,
    this.biometricButton = const Icon(Icons.fingerprint),
    this.onCompleted,
    this.canBiometric = false,
    this.showBiometricFirst = false,
    this.biometricAuthenticate,
    this.showBiometricFirstController,
    this.backgroundColor = Colors.white,
    this.backgroundColorOpacity = 0.5,
    this.onUnlocked,
    this.maxRetries = -1,
    this.didMaxRetries,
    this.onError,
  });

  @override
  _LockScreenState createState() => _LockScreenState();
}

class _LockScreenState extends State<LockScreen> {
  // receive from circle input button
  final StreamController<String> enteredStream = StreamController<String>();
  final StreamController<void> removedStreamController =
      StreamController<void>();
  final StreamController<int> enteredLengthStream =
      StreamController<int>.broadcast();
  final StreamController<bool> validateStreamController =
      StreamController<bool>();

  // control for Android back button
  bool _needClose = false;

  // confirm flag
  bool _isConfirmation = false;

  // confirm verify passcode
  String _verifyConfirmPasscode = '';

  List<String> enteredValues = <String>[];

  int _retries = 0;

  @override
  void initState() {
    super.initState();

    SystemChrome.setPreferredOrientations([
      DeviceOrientation.portraitUp,
      DeviceOrientation.portraitDown,
    ]);

    if (widget.showBiometricFirst) {
      if (widget.biometricAuthenticate != null) {
        // Set the listener if there is a stream option.
        if (widget.showBiometricFirstController != null) {
          widget.showBiometricFirstController?.stream.listen((_) {
            widget.biometricAuthenticate!(context).then((unlocked) {
              if (unlocked) {
                Navigator.of(context).pop();

                if (widget.onUnlocked != null) {
                  widget.onUnlocked!();
                }
              }
            });
          });
        } else {
          // It is executed by a certain time.
          Future.delayed(
            Duration(milliseconds: 350),
            () {
              widget.biometricAuthenticate!(context).then((unlocked) {
                if (unlocked) {
                  Navigator.pop(context);

                  if (widget.onUnlocked != null) {
                    widget.onUnlocked!();
                  }
                }
              });
            },
          );
        }
      }
    }
  }

  void _removedStreamListener() {
    if (removedStreamController.hasListener) {
      return;
    }

    removedStreamController.stream.listen((_) {
      if (enteredValues.isNotEmpty) {
        enteredValues.removeLast();
      }

      enteredLengthStream.add(enteredValues.length);
    });
  }

  void _enteredStreamListener() {
    if (enteredStream.hasListener) {
      return;
    }

    enteredStream.stream.listen((value) {
      // add list entered value
      enteredValues.add(value);
      enteredLengthStream.add(enteredValues.length);

      // the same number of digits was entered.
      if (enteredValues.length == widget.digits) {
        var buffer = StringBuffer();
        enteredValues.forEach((value) {
          buffer.write(value);
        });
        _verifyCorrectString(buffer.toString());
      }
    });
  }

  void _verifyCorrectString(String enteredValue) {
    Future.delayed(Duration(milliseconds: 150), () {
      var _verifyPasscode = widget.correctString;

      if (widget.confirmMode) {
        if (_isConfirmation == false) {
          _verifyConfirmPasscode = enteredValue;
          enteredValues.clear();
          enteredLengthStream.add(enteredValues.length);
          _isConfirmation = true;
          setState(() {});
          return;
        }
        _verifyPasscode = _verifyConfirmPasscode;
      }

      if (enteredValue == _verifyPasscode) {
        // send valid status to DotSecretUI
        validateStreamController.add(true);
        enteredValues.clear();
        enteredLengthStream.add(enteredValues.length);

        if (widget.onCompleted != null) {
          // call user function
          widget.onCompleted!(context, enteredValue);
        } else {
          _needClose = true;
          Navigator.pop(context);
        }

        if (widget.onUnlocked != null) {
          widget.onUnlocked!();
        }
      } else {
        // send invalid status to DotSecretUI
        validateStreamController.add(false);
        enteredValues.clear();
        enteredLengthStream.add(enteredValues.length);
        _verifyMaxRetries();
        _incorrect(_retries);
      }
    });
  }

  /// Check if the maximum number of retries has been reached.
  void _verifyMaxRetries() {
    if (_retries >= widget.maxRetries) {
      if (widget.didMaxRetries != null) {
        widget.didMaxRetries!();
      }
    }

    _retries++;
  }

  void _incorrect(int retries) {
    if (widget.onError != null) {
      widget.onError!(retries);
    }
  }

  @override
  Widget build(BuildContext context) {
    _enteredStreamListener();
    _removedStreamListener();
    var _rowMarginSize = MediaQuery.of(context).size.width * 0.025;
    var _columnMarginSize = MediaQuery.of(context).size.width * 0.065;

    return WillPopScope(
      onWillPop: () async {
        if (_needClose || widget.canCancel) {
          return true;
        }

        return false;
      },
      child: Scaffold(
        backgroundColor:
            widget.backgroundColor.withOpacity(widget.backgroundColorOpacity),
        body: SafeArea(
          child: BackdropFilter(
            filter: ImageFilter.blur(sigmaX: 3.5, sigmaY: 3.5),
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                _buildTitle(),
                DotSecretUI(
                  validateStream: validateStreamController.stream,
                  dots: widget.digits,
                  config: widget.dotSecretConfig,
                  enteredLengthStream: enteredLengthStream.stream,
                ),
                Padding(
                  padding: EdgeInsets.symmetric(
                    horizontal: _columnMarginSize,
                  ),
                  child: Column(
                    children: <Widget>[
                      Container(
                        margin: EdgeInsets.symmetric(vertical: _rowMarginSize),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                          children: <Widget>[
                            _buildNumberTextButton(context, '1'),
                            _buildNumberTextButton(context, '2'),
                            _buildNumberTextButton(context, '3'),
                          ],
                        ),
                      ),
                      Container(
                        margin: EdgeInsets.symmetric(vertical: _rowMarginSize),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                          children: <Widget>[
                            _buildNumberTextButton(context, '4'),
                            _buildNumberTextButton(context, '5'),
                            _buildNumberTextButton(context, '6'),
                          ],
                        ),
                      ),
                      Container(
                        margin: EdgeInsets.symmetric(vertical: _rowMarginSize),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                          children: <Widget>[
                            _buildNumberTextButton(context, '7'),
                            _buildNumberTextButton(context, '8'),
                            _buildNumberTextButton(context, '9'),
                          ],
                        ),
                      ),
                      Container(
                        margin: EdgeInsets.symmetric(vertical: _rowMarginSize),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                          children: <Widget>[
                            _buildBothSidesButton(context, _biometricButton()),
                            _buildNumberTextButton(context, '0'),
                            _buildBothSidesButton(context, _rightSideButton()),
                          ],
                        ),
                      ),
                    ],
                  ),
                )
              ],
            ),
          ),
        ),
      ),
    );
  }

  Widget _buildNumberTextButton(
    BuildContext context,
    String number,
  ) {
    final buttonSize = MediaQuery.of(context).size.width * 0.215;
    return Container(
      width: buttonSize,
      height: buttonSize,
      child: CircleInputButton(
        enteredSink: enteredStream.sink,
        text: number,
        config: widget.circleInputButtonConfig,
      ),
    );
  }

  Widget _buildBothSidesButton(BuildContext context, Widget button) {
    final buttonSize = MediaQuery.of(context).size.width * 0.215;
    return Container(
      width: buttonSize,
      height: buttonSize,
      child: button,
    );
  }

  Widget _buildTitle() {
    return Container(
      margin: EdgeInsets.symmetric(vertical: 20),
      child: Text(
        _isConfirmation ? widget.confirmTitle : widget.title,
        style: TextStyle(fontSize: 20.0),
      ),
    );
  }

  Widget _biometricButton() {
    if (!widget.canBiometric) return Container();

    return OutlinedButton(
      child: widget.biometricButton,
      onPressed: () {
        if (widget.biometricAuthenticate != null) {
          widget.biometricAuthenticate!(context).then((unlocked) {
            if (unlocked) {
              Navigator.pop(context);

              if (widget.onUnlocked != null) {
                widget.onUnlocked!();
              }
            }
          });
        }
      },
      style: OutlinedButton.styleFrom(
        side: BorderSide.none,
        shape: CircleBorder(),
      ),
    );
  }

  Widget _rightSideButton() {
    if (widget.rightSideButton != null) {
      return widget.rightSideButton!;
    }

    return StreamBuilder<int>(
        stream: enteredLengthStream.stream,
        builder: (context, snapshot) {
          String buttonText;
          if (snapshot.hasData && snapshot.data! > 0) {
            buttonText = widget.deleteText!;
          } else if (widget.canCancel) {
            buttonText = widget.cancelText!;
          } else {
            return Container();
          }

          return TextButton(
            child: Text(
              buttonText,
              style: TextStyle(
                fontSize: MediaQuery.of(context).size.width * 0.055,
              ),
              softWrap: false,
              textAlign: TextAlign.center,
            ),
            onPressed: () {
              if (snapshot.hasData && snapshot.data! > 0) {
                removedStreamController.sink.add(null);
              } else {
                if (widget.canCancel) {
                  _needClose = true;
                  Navigator.of(context).maybePop();
                }
              }
            },
            style: TextButton.styleFrom(
              shape: CircleBorder(),
              backgroundColor: Colors.transparent,
            ),
          );
        });
  }

  @override
  void dispose() {
    enteredStream.close();
    enteredLengthStream.close();
    validateStreamController.close();
    removedStreamController.close();
    if (widget.showBiometricFirstController != null) {
      widget.showBiometricFirstController!.close();
    }

    // restore orientation.
    SystemChrome.setPreferredOrientations([
      DeviceOrientation.landscapeLeft,
      DeviceOrientation.landscapeRight,
      DeviceOrientation.portraitDown,
      DeviceOrientation.portraitUp,
    ]);

    super.dispose();
  }
}


example/lib/main.dart
import 'package:example/second_page.dart';
import 'package:flutter/material.dart';
import 'package:flutter_screen_lock/circle_input_button.dart';
import 'package:flutter_screen_lock/lock_screen.dart';
import 'package:local_auth/local_auth.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({required this.title});
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              ElevatedButton(
                child: Text('Open Lock Screen'),
                onPressed: () => showLockScreen(
                  context: context,
                  correctString: '1234',
                  onCompleted: (context, result) {
                    // if you specify this callback,
                    // you must close the screen yourself
                    Navigator.of(context).maybePop();
                  },
                  onUnlocked: () => print('Unlocked.'),
                ),
              ),
              ElevatedButton(
                child: Text('6 Digits'),
                onPressed: () => showLockScreen(
                  context: context,
                  digits: 6,
                  correctString: '123456',
                ),
              ),
              ElevatedButton(
                child: Text('Use local_auth'),
                onPressed: () => showLockScreen(
                  context: context,
                  correctString: '1234',
                  canBiometric: true,
                  // biometricButton is default Icon(Icons.fingerprint)
                  // When you want to change the icon with `BiometricType.face`, etc.
                  biometricButton: Icon(Icons.face),
                  biometricAuthenticate: (context) async {
                    final localAuth = LocalAuthentication();
                    final didAuthenticate = await localAuth.authenticate(
                      localizedReason: 'Please authenticate',
                      biometricOnly: true,
                    );

                    if (didAuthenticate) {
                      return true;
                    }

                    return false;
                  },
                  onUnlocked: () {
                    print('Unlocked.');
                  },
                ),
              ),
              ElevatedButton(
                child: Text('Open biometric first'),
                onPressed: () => showLockScreen(
                  context: context,
                  correctString: '1234',
                  canBiometric: true,
                  showBiometricFirst: true,
                  biometricAuthenticate: (_) async {
                    final localAuth = LocalAuthentication();
                    final didAuthenticate = await localAuth.authenticate(
                      localizedReason: 'Please authenticate',
                      biometricOnly: true,
                    );

                    if (didAuthenticate) {
                      return true;
                    }

                    return false;
                  },
                  onUnlocked: () => print('Unlocked.'),
                ),
              ),
              ElevatedButton(
                child: Text('Go to another page after unlocked biometrics'),
                onPressed: () => showLockScreen(
                  context: context,
                  correctString: '1234',
                  canBiometric: true,
                  showBiometricFirst: true,
                  biometricAuthenticate: (_) async {
                    final localAuth = LocalAuthentication();
                    final didAuthenticate = await localAuth.authenticate(
                      localizedReason: 'Please authenticate',
                      biometricOnly: true,
                    );

                    if (didAuthenticate) {
                      return true;
                    }

                    return false;
                  },
                  onUnlocked: () async => Navigator.push(
                    context,
                    MaterialPageRoute(
                      builder: (context) => SecondPage(),
                    ),
                  ),
                ),
              ),
              ElevatedButton(
                child: Text('Can\'t cancel'),
                onPressed: () => showLockScreen(
                  context: context,
                  correctString: '1234',
                  canCancel: false,
                ),
              ),
              ElevatedButton(
                child: Text('Customize text'),
                onPressed: () => showLockScreen(
                  context: context,
                  correctString: '1234',
                  cancelText: 'Close',
                  deleteText: 'Remove',
                ),
              ),
              ElevatedButton(
                child: Text('Confirm mode.'),
                onPressed: () => showConfirmPasscode(
                  context: context,
                  onCompleted: (context, verifyCode) {
                    print(verifyCode);
                    Navigator.of(context).maybePop();
                  },
                ),
              ),
              ElevatedButton(
                child: Text('Change styles.'),
                onPressed: () => showLockScreen(
                  context: context,
                  correctString: '1234',
                  backgroundColor: Colors.grey.shade50,
                  backgroundColorOpacity: 1,
                  circleInputButtonConfig: CircleInputButtonConfig(
                    textStyle: TextStyle(
                      fontSize: MediaQuery.of(context).size.width * 0.1,
                      color: Colors.white,
                    ),
                    backgroundColor: Colors.blue,
                    backgroundOpacity: 0.5,
                    shape: RoundedRectangleBorder(
                      side: BorderSide(
                        width: 1,
                        color: Colors.blue,
                        style: BorderStyle.solid,
                      ),
                    ),
                  ),
                ),
              ),
              ElevatedButton(
                child: Text('Max retries / Incorrect event'),
                onPressed: () => showLockScreen(
                  context: context,
                  correctString: '1234',
                  maxRetries: 1,
                  onError: (retries) {
                    print(retries);
                  },
                  didMaxRetries: () {
                    Navigator.pop(context);
                    showAboutDialog(
                      context: context,
                      children: [
                        Text('The maximum number of retries has been reached.'),
                      ],
                    );
                  },
                  biometricAuthenticate: (_) async {
                    final localAuth = LocalAuthentication();
                    final didAuthenticate = await localAuth.authenticate(
                      localizedReason: 'Please authenticate',
                      biometricOnly: true,
                    );

                    if (didAuthenticate) {
                      return true;
                    }

                    return false;
                  },
                  canBiometric: true,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
